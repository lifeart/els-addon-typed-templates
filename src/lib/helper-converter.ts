const ts = require('ts-morph');

const Project = ts.Project;
const tNode = ts.Node;

export function getHelper(filePath) {

  const project = new Project({ compilerOptions: { allowJs: true } });

  const helperFile = project.addSourceFileAtPath(filePath);

  const defaultExportSymbol = helperFile.getDefaultExportSymbol();

  if (defaultExportSymbol) {
    const declaration = defaultExportSymbol.getDeclarations()[0];
    if (declaration && tNode.isExportAssignment(declaration)) {
      let helperWrapper = declaration.getExpression();
      let helper;
      let helperWrapperDefinition;
      if (tNode.isIdentifier(helperWrapper)) {
        helperWrapperDefinition = helperWrapper.getDefinitionNodes()[0];
        helperWrapper = helperWrapperDefinition.getChildren().slice(-1)[0];
      }

      helper = helperWrapper.getArguments()[0];
      if (tNode.isIdentifier(helper)) {
        helper = helper.getDefinitionNodes()[0];
      }

      if (helper.isExported && helper.isExported()) {
        helper.setIsExported(false);
      }

      const txt = helper.getFullText();
      helper.remove();
      helperWrapperDefinition?.remove();
      declaration.remove();

      return `${helperFile.getFullText()} export default ${txt}`;
    }
  }
  return null;
}
