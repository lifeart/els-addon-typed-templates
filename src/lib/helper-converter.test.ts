import { getHelper } from './helper-converter';
import * as path from 'path';

describe('getHelper', () => {
    it('it works', ()=>{
        expect(getHelper(path.join(__dirname, './../../app/helpers/filter-attrs.ts'))).toMatchSnapshot();
    })
});