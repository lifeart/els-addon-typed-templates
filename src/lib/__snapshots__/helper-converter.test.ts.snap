// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getHelper it works 1`] = `null`;
